// @m4ykeldev/notebooklm-mcp/dist/chunk-NYG5WIAB.js
#!/usr/bin/env node

// src/auth.ts
import { homedir } from "os";
import { join } from "path";
import { readFileSync, writeFileSync, mkdirSync, existsSync } from "fs";
import { execSync } from "child_process";
import { createInterface } from "readline";

// src/constants.ts
var CodeMapper = class {
  nameToCode;
  codeToName;
  unknownLabel;
  displayNames;
  constructor(mapping, unknownLabel = "unknown") {
    this.nameToCode = new Map(
      Object.entries(mapping).map(([k, v]) => [k.toLowerCase(), v])
    );
    this.codeToName = new Map(
      Object.entries(mapping).map(([k, v]) => [v, k])
    );
    this.unknownLabel = unknownLabel;
    this.displayNames = Object.keys(mapping).sort();
  }
  getCode(name) {
    const code = this.nameToCode.get(name.toLowerCase());
    if (code === void 0) {
      throw new Error(
        `Invalid value "${name}". Valid options: ${this.optionsStr()}`
      );
    }
    return code;
  }
  getName(code) {
    if (code === null) return this.unknownLabel;
    return this.codeToName.get(code) ?? this.unknownLabel;
  }
  optionsStr() {
    return this.displayNames.join(", ");
  }
  names() {
    return [...this.displayNames];
  }
};
var RPC_IDS = {
  LIST_NOTEBOOKS: "wXbhsf",
  GET_NOTEBOOK: "rLM1Ne",
  CREATE_NOTEBOOK: "CCqFvf",
  RENAME_NOTEBOOK: "s0tc2d",
  DELETE_NOTEBOOK: "WWINqb",
  ADD_SOURCE: "izAoDd",
  GET_SOURCE: "hizoJc",
  CHECK_FRESHNESS: "yR9Yof",
  SYNC_DRIVE: "FLmJqe",
  DELETE_SOURCE: "tGMBJ",
  QUERY: "ZAnZ8",
  GET_CONVERSATIONS: "hPTbtc",
  PREFERENCES: "hT54vc",
  SUBSCRIPTION: "ozz5Z",
  SETTINGS: "ZwVcOc",
  GET_SUMMARY: "VfAZjd",
  GET_SOURCE_GUIDE: "tr032e",
  START_FAST_RESEARCH: "Ljjv0c",
  START_DEEP_RESEARCH: "QA9ei",
  POLL_RESEARCH: "e3bVqc",
  IMPORT_RESEARCH: "LBwxtb",
  CREATE_STUDIO: "R7cb6c",
  POLL_STUDIO: "gArtLc",
  DELETE_STUDIO: "V5N4be",
  GENERATE_MIND_MAP: "yyryJe",
  SAVE_MIND_MAP: "CYK0Xb",
  LIST_MIND_MAPS: "cFji9",
  DELETE_MIND_MAP: "AH0mwd"
};
var OWNERSHIP_MINE = 1;
var CHAT_GOALS = new CodeMapper({
  default: 1,
  custom: 2,
  learning_guide: 3
});
var CHAT_RESPONSE_LENGTHS = new CodeMapper({
  default: 1,
  longer: 4,
  shorter: 5
});
var RESEARCH_SOURCES = new CodeMapper({ web: 1, drive: 2 });
var RESEARCH_MODES = new CodeMapper({ fast: 1, deep: 5 });
var RESULT_TYPES = new CodeMapper({
  web: 1,
  google_doc: 2,
  google_slides: 3,
  deep_report: 5,
  google_sheets: 8
});
var SOURCE_TYPES = new CodeMapper({
  google_docs: 1,
  google_slides_sheets: 2,
  pdf: 3,
  pasted_text: 4,
  web_page: 5,
  generated_text: 8,
  youtube: 9,
  uploaded_file: 11,
  image: 13,
  word_doc: 14
});
var STUDIO_TYPES = new CodeMapper({
  audio: 1,
  report: 2,
  video: 3,
  flashcards: 4,
  infographic: 7,
  slide_deck: 8,
  data_table: 9
});
var AUDIO_FORMATS = new CodeMapper({
  deep_dive: 1,
  brief: 2,
  critique: 3,
  debate: 4
});
var AUDIO_LENGTHS = new CodeMapper({
  short: 1,
  default: 2,
  long: 3
});
var VIDEO_FORMATS = new CodeMapper({ explainer: 1, brief: 2 });
var VIDEO_STYLES = new CodeMapper({
  auto_select: 1,
  custom: 2,
  classic: 3,
  whiteboard: 4,
  kawaii: 5,
  anime: 6,
  watercolor: 7,
  retro_print: 8,
  heritage: 9,
  paper_craft: 10
});
var INFOGRAPHIC_ORIENTATIONS = new CodeMapper({
  landscape: 1,
  portrait: 2,
  square: 3
});
var INFOGRAPHIC_DETAILS = new CodeMapper({
  concise: 1,
  standard: 2,
  detailed: 3
});
var SLIDE_DECK_FORMATS = new CodeMapper({
  detailed_deck: 1,
  presenter_slides: 2
});
var SLIDE_DECK_LENGTHS = new CodeMapper({
  short: 1,
  default: 3
});
var FLASHCARD_DIFFICULTIES = new CodeMapper({
  easy: 1,
  medium: 2,
  hard: 3
});
var FLASHCARD_COUNT_DEFAULT = 2;
var REPORT_FORMATS = {
  "Briefing Doc": {
    title: "Briefing Doc",
    description: "A comprehensive briefing document",
    prompt: "Create a briefing document"
  },
  "Study Guide": {
    title: "Study Guide",
    description: "A study guide for the material",
    prompt: "Create a study guide"
  },
  "Blog Post": {
    title: "Blog Post",
    description: "A blog post about the material",
    prompt: "Create a blog post"
  },
  "Create Your Own": {
    title: "Create Your Own",
    description: "Custom report format",
    prompt: ""
  }
};
var BASE_URL = "https://notebooklm.google.com";
var BATCHEXECUTE_PATH = "/_/LabsTailwindUi/data/batchexecute";
var QUERY_PATH = "/_/LabsTailwindUi/data/google.internal.labs.tailwind.orchestration.v1.LabsTailwindOrchestrationService/GenerateFreeFormStreamed";
var DEFAULT_BL = "boq_labs-tailwind-frontend_20260218.06_p0";
var REQUIRED_COOKIES = [
  "SID",
  "HSID",
  "SSID",
  "APISID",
  "SAPISID"
];
var USER_AGENT = "Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/133.0.0.0 Safari/537.36";
var DEFAULT_TIMEOUT = 3e4;
var EXTENDED_TIMEOUT = 12e4;

// src/auth.ts
var CONFIG_DIR = join(homedir(), ".notebooklm-mcp");
var AUTH_FILE = join(CONFIG_DIR, "auth.json");
var CHROME_PROFILE = join(CONFIG_DIR, "chrome-profile");
function validateCookies(cookies) {
  return REQUIRED_COOKIES.every((name) => name in cookies);
}
function buildCookieHeader(cookies) {
  return Object.entries(cookies).map(([k, v]) => `${k}=${v}`).join("; ");
}
function extractCsrfFromPage(html) {
  const patterns = [
    /"SNlM0e":"([^"]+)"/,
    /at=([^&"]+)/,
    /"FdrFJe":"([^"]+)"/
  ];
  for (const pattern of patterns) {
    const match = html.match(pattern);
    if (match) return match[1];
  }
  return null;
}
function extractSessionIdFromPage(html) {
  const patterns = [/"FdrFJe":"([^"]+)"/, /f\.sid=(\d+)/];
  for (const pattern of patterns) {
    const match = html.match(pattern);
    if (match) return match[1];
  }
  return null;
}
function saveTokens(tokens) {
  mkdirSync(CONFIG_DIR, { recursive: true });
  writeFileSync(AUTH_FILE, JSON.stringify(tokens, null, 2), "utf-8");
}
function loadTokensFromCache() {
  if (!existsSync(AUTH_FILE)) return null;
  try {
    const data = JSON.parse(readFileSync(AUTH_FILE, "utf-8"));
    if (data.cookies && validateCookies(data.cookies)) {
      return {
        cookies: data.cookies,
        csrf_token: data.csrf_token || "",
        session_id: data.session_id || "",
        extracted_at: data.extracted_at || 0
      };
    }
    return null;
  } catch {
    return null;
  }
}
function loadTokensFromEnv() {
  const cookieStr = process.env.NOTEBOOKLM_COOKIES;
  if (!cookieStr) return null;
  const cookies = {};
  for (const part of cookieStr.split(";")) {
    const eq = part.indexOf("=");
    if (eq > 0) {
      cookies[part.slice(0, eq).trim()] = part.slice(eq + 1).trim();
    }
  }
  if (!validateCookies(cookies)) return null;
  return {
    cookies,
    csrf_token: process.env.NOTEBOOKLM_CSRF_TOKEN || "",
    session_id: process.env.NOTEBOOKLM_SESSION_ID || "",
    extracted_at: Date.now() / 1e3
  };
}
function loadTokens() {
  const fromEnv = loadTokensFromEnv();
  if (fromEnv) return fromEnv;
  const fromCache = loadTokensFromCache();
  if (fromCache) return fromCache;
  throw new Error(
    "No authentication tokens found. Run `npx @m4ykeldev/notebooklm-mcp auth` to authenticate, or set NOTEBOOKLM_COOKIES environment variable."
  );
}
function parseCookieString(raw) {
  const cookies = {};
  for (const line of raw.split("\n")) {
    const trimmed = line.trim();
    if (!trimmed || trimmed.startsWith("#")) continue;
    for (const part of trimmed.split(";")) {
      const eq = part.indexOf("=");
      if (eq > 0) {
        cookies[part.slice(0, eq).trim()] = part.slice(eq + 1).trim();
      }
    }
  }
  return cookies;
}
async function runFileImport(filePath) {
  if (!filePath) {
    console.log(`
To authenticate via file:
1. Open Chrome and navigate to https://notebooklm.google.com
2. Open DevTools (F12) > Network tab
3. Type "batchexecute" in the filter
4. Click on any batchexecute request
5. Find "cookie:" in Request Headers
6. Copy the full cookie VALUE (not the header name)
7. Save to a file and provide the path
`);
    throw new Error("Provide a cookie file path with --file <path>");
  }
  const raw = readFileSync(filePath, "utf-8");
  const cookies = parseCookieString(raw);
  if (!validateCookies(cookies)) {
    throw new Error(
      `Missing required cookies. Need: ${REQUIRED_COOKIES.join(", ")}`
    );
  }
  const tokens = {
    cookies,
    csrf_token: "",
    session_id: "",
    extracted_at: Date.now() / 1e3
  };
  saveTokens(tokens);
  console.log("Authentication tokens saved successfully.");
  return tokens;
}
function readLineFromStdin(prompt) {
  return new Promise((resolve) => {
    const rl = createInterface({
      input: process.stdin,
      output: process.stdout
    });
    rl.question(prompt, (answer) => {
      rl.close();
      resolve(answer.trim());
    });
  });
}
function openInBrowser(url) {
  const platform = process.platform;
  try {
    if (platform === "linux") {
      execSync(`xdg-open "${url}"`, { stdio: "ignore" });
    } else if (platform === "darwin") {
      execSync(`open "${url}"`, { stdio: "ignore" });
    } else if (platform === "win32") {
      execSync(`start "" "${url}"`, { stdio: "ignore" });
    }
  } catch {
    console.log(`Could not open browser automatically. Open this URL manually:
${url}`);
  }
}
async function runAuthFlow() {
  console.log("\u2554\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2557");
  console.log("\u2551         NotebookLM MCP \u2014 Authentication Setup          \u2551");
  console.log("\u255A\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u2550\u255D\n");
  console.log("Opening NotebookLM in your browser...\n");
  openInBrowser(BASE_URL);
  console.log("Follow these steps:\n");
  console.log("  1. NotebookLM should open in your browser (you're already logged in)");
  console.log("  2. Press F12 to open DevTools");
  console.log("  3. Go to the Network tab");
  console.log("  4. Type 'batchexecute' in the filter box");
  console.log("  5. Click on any request that appears in the list");
  console.log("  6. In the Headers panel, find 'cookie:' under Request Headers");
  console.log("  7. Right-click the cookie value \u2192 Copy value\n");
  console.log("  Tip: If no requests appear, refresh the page (F5) with DevTools open.\n");
  const cookieStr = await readLineFromStdin("Paste the cookie value here: ");
  if (!cookieStr) {
    throw new Error("No cookie string provided.");
  }
  const cookies = parseCookieString(cookieStr);
  if (!validateCookies(cookies)) {
    console.log("\n\u274C Missing required cookies.");
    console.log(`   Need: ${REQUIRED_COOKIES.join(", ")}`);
    console.log(`   Got:  ${Object.keys(cookies).join(", ")}`);
    throw new Error("Invalid cookie string. Make sure you copied the full cookie value.");
  }
  const tokens = {
    cookies,
    csrf_token: "",
    session_id: "",
    extracted_at: Date.now() / 1e3
  };
  saveTokens(tokens);
  console.log(`
\u2705 Authentication saved successfully!`);
  console.log(`   ${Object.keys(cookies).length} cookies extracted`);
  console.log(`   Stored in: ~/.notebooklm-mcp/auth.json`);
  console.log(`
   CSRF token and session ID will be auto-extracted on first use.`);
  return tokens;
}
function showTokens() {
  const tokens = loadTokensFromCache();
  if (!tokens) {
    console.log("No cached tokens found.");
    return;
  }
  const cookieNames = Object.keys(tokens.cookies);
  const hasRequired = REQUIRED_COOKIES.every((c) => cookieNames.includes(c));
  const age = tokens.extracted_at ? Math.round((Date.now() / 1e3 - tokens.extracted_at) / 3600) : "unknown";
  console.log(`Cached tokens:`);
  console.log(`  Cookies: ${cookieNames.length} (${cookieNames.join(", ")})`);
  console.log(`  Required cookies present: ${hasRequired ? "yes" : "NO"}`);
  console.log(`  CSRF token: ${tokens.csrf_token ? "present" : "missing"}`);
  console.log(`  Session ID: ${tokens.session_id ? "present" : "missing"}`);
  console.log(`  Age: ${age} hours`);
  console.log(`  File: ${AUTH_FILE}`);
}

export {
  RPC_IDS,
  OWNERSHIP_MINE,
  CHAT_GOALS,
  CHAT_RESPONSE_LENGTHS,
  RESEARCH_SOURCES,
  RESEARCH_MODES,
  RESULT_TYPES,
  SOURCE_TYPES,
  STUDIO_TYPES,
  AUDIO_FORMATS,
  AUDIO_LENGTHS,
  VIDEO_FORMATS,
  VIDEO_STYLES,
  INFOGRAPHIC_ORIENTATIONS,
  INFOGRAPHIC_DETAILS,
  SLIDE_DECK_FORMATS,
  SLIDE_DECK_LENGTHS,
  FLASHCARD_DIFFICULTIES,
  FLASHCARD_COUNT_DEFAULT,
  REPORT_FORMATS,
  BASE_URL,
  BATCHEXECUTE_PATH,
  QUERY_PATH,
  DEFAULT_BL,
  USER_AGENT,
  DEFAULT_TIMEOUT,
  EXTENDED_TIMEOUT,
  validateCookies,
  buildCookieHeader,
  extractCsrfFromPage,
  extractSessionIdFromPage,
  saveTokens,
  loadTokensFromCache,
  loadTokensFromEnv,
  loadTokens,
  runFileImport,
  runAuthFlow,
  showTokens
};
